d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -211,8 +219,102 @@ public void onSuccess(None result)
         }
         runNextMarkUpOrMarkDown();
       }
-    });
+    };
+
+    final Callback<None> doMarkUpCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e)
+      {
+        // this would ideally never happen
+        callback.onError(e);
+      }
+
+      @Override
+      public void onSuccess(None result)
+      {
+        _log.info("doing mark up of main cluster " + _cluster);
+        _server.markUp(_cluster, _uri, _partitionDataMap, _uriSpecificProperties, markUpCallback);
+      }
+    };
+
+    final Callback<None> warmupMarkDownCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e) {
+        _log.info("mark down warm up cluster " + _warmupClusterName + " has failed");
+        // what to do here ? - go ahead continue to announce main cluster
+        // doAnnounceCallback.onSuccess(None.none());
+      }
+
+      @Override
+      public void onSuccess(None result) {
+        _log.info("mark down warm up cluster " + _warmupClusterName +
+            " has completed, now mark up of main cluster " + _cluster);
+        doMarkUpCallback.onSuccess(None.none());
+      }
+    };
+
+    final Callback<None> doWarmupMarkDownCallback = new Callback<None>() {
+      @Override
+      public void onError(Throwable e) {
+        _log.info("failure in scheduling warm up mark down");
+        // what to do here ? - go ahead continue to announce main cluster
```

Comment:
imo, there are a couple of things we can do:
1. Notice that in this class we also have a `ZKPersistentConnection.EventListener` that subscribes to zk events. I would consider recording a failed markdown, and retry the operation whenever zk connection is back.
2. As you mentioned, we should go ahead and continue to announce the main cluster

Reply:
Thank you for suggestions and explaining that failed markdown for warm up should be retried. About announcing to main cluster, in the markUp method of ZooKeeperServer, in the getCallback the check for uri already existing will not be able to catch that the uri exists but for a different cluster name (warm up cluster's name) and will add association of the main cluster name with the uri in the store, in this case should we still announce to the main cluster or keep retrying mark down for the warm up cluster?https://github.com/linkedin/rest.li/blob/7372bc76f7ce78007ecb24ca80f17399e5246e13/d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperServer.java#L145

Reply:
Updated the design to retry markDown for warm-up cluster in case of ZooKeeper connection loss or session expiry. In case of other exceptions will continue to announce to main cluster.